### ISSUE TITLE
Causing player selector in Minecraft command

### ISSUE THREAD

**Issue by nguyenquyhy** [SpongePowered/Sponge#874]
Currently Minecraft supports several target selectors such as `@p`, `@r`, `@a`. I wonder if Sponge can add a new selector to target the cause of the command.

For example, if I put a button on a command block, and set the command block to `/tp @c 100 100 100`, the command block will teleport the exact player that pressed the button. At the moment, the best I can do is to set `tp @p 100 100 100`, which is quite problematic in crowded area.

If Sponge can really override the player target mechanism of Minecraft, the same `@c` may also be used with other plugins such as Command Sign.

**Comment by ryantheleach** [comment]
Is it possible to implement this in a plugin?

**Comment by nguyenquyhy** [comment]
An API that exposes Minecraft target resolution (for those new @x) would be great. However, since cause tracking is supported by Sponge, I think at least a cause target selector should be something built-in.

**Comment by gabizou** [comment]
@kenzierocks It is within the realm of possibility, but I'm curious how it would be handled from the API standpoint. I'd imagine it would still be a special `CommandElement`, and @zml2008 did express some interest in such a thing. Implementation wise, it would be really simple with the `CauseTracker` system in place for both 1.8.9 and 1.10.2

**Comment by nguyenquyhy** [comment]
I think on the API side, a new event or changing existing event might not be necessary. However, internal Sponge's resolution for player parameters (those that declared during command registration) can be changed to make use of this `@c` feature as well. Moreover, it would be great to add a function like resolveEntityTarget() which accepts those `@` string.

**Comment by kenzierocks** [comment]
The SelectorResolver does that, but it has a ton of bugs that need fixing (I have an open PR for this).

**Comment by JBYoshi** [comment]
@nguyenquyhy I added support for that some time ago in SpongePowered/SpongeAPI#780, but it does use SelectorResolver, so it will be affected by any bugs in that (see kashike's comment).

**Comment by nguyenquyhy** [comment]
@JBYoshi it seems the PR you referenced is closed. Is it merged somewhere else in Sponge?

**Comment by JBYoshi** [comment]
It was rebased into the master branch in https://github.com/SpongePowered/SpongeAPI/commit/347605274bf3d3eae9ece79a3df7f23d26628ff8.

**Comment by nguyenquyhy** [comment]
I don't really get it. So how can I target causing player now? E.g. a player clicks a button on a command block and how can the command block do `/tell ...` to that exact player?

**Comment by JBYoshi** [comment]
I meant that added support for custom selectors in commands. It does not add support for a cause selector.

**Comment by advancid** [comment]
How would it works if there is a button and a complex redstone circuit with gate, ...?
I'm not sure it would be possible.

**Comment by ryantheleach** [comment]
I too thought it wouldn't be possible at the start, then when the Sponge cause system was introduced I was pleasantly surprised.

At this stage it's just creating a selector that responds to causes, the cause system already exists, and works.

**Comment by nguyenquyhy** [comment]
Yes that is why I give the suggestion. It seems to me that Sponge already has all neccessary facility to implement a cause selector (Selector resolver + Cause tracking). @c can become a new built-in selector for all commands that use selector resolver.

**Comment by nguyenquyhy** [comment]
@advancid since the feature relies on cause tracker, it will have all limitation/ability of cause tracker. 

I agree that this feature may not work accurately in complex redstone mechanics when the cause is not obvious. However, I think for most of the cases, the cause of an activation is quite obvious, and the lack of cause selector force server owner to employ complex mechanism to make sure they target the correct player. Moreover, with plugins, there is also fewer needs for server owners to build complex redstone mechanics for complex functions.

**Comment by ryantheleach** [comment]
The only thing I don't necessarily agree with is using the prefix @ for Sponge-custom selectors, as it's more plausible that minecraft introducing additional selectors would also use the @ prefix.

It's a minor gripe, but I could see it being a future migration pain point if ``@c`` were ever to be used by minecraft itself.

**Comment by kenzierocks** [comment]
Would it be too much of a pain to use `@sponge:c`, identifying like items and blocks do?

**Comment by nguyenquyhy** [comment]
For me, `@c` or `@sponge:c` or anything else is ok as long as it is implemented.

**Comment by advancid** [comment]
Sponge is "A community-driven open source Minecraft modding platform.", I don't think this should be imlemented in Sponge it self but maybe in a plugin.

**Comment by nguyenquyhy** [comment]
I believe Sponge already has all facility to implement the function. In my opinion, having a another plugin just to merge two built-in Sponge functions feel a bit overkill. 
I am aware that there can be compatibility issue when Minecraft moves forward. That's why I really hope to hear from Sponge team to know if they will go ahead with the implementation or I should build something on my own. This has been here for more than 2 weeks already.

**Comment by nguyenquyhy** [comment]
Is there any update on this?

**Comment by m0pt0pmatt** [comment]
I also would be interested in seeing this added to Sponge by API 5.

**Comment by advancid** [comment]
If this was implemented, I think it would not be in the API since it would make no sense, the API is for plugins.

**Comment by ST-DDT** [comment]
@advancid It might be a useful default for some commands.

`/tpto someplace` (Teleport executor)
`/tpto someplace @a` (Teleport all)

**Comment by mattysweeps** [comment]
Sorry to wake an old issue. Does anyone still want this? 

Sponge Causes have come a long way since this was submitted. Does anyone have a better idea of whether or not it is feasible? Or a better idea whether or not it belongs as a part of Sponge?
